### PR TITLE
Implementation of initial commands and 'fish' invocation tests

### PR DESCRIPTION
**Pull request by gerph**
## Description

This change implements the '--init-command' or '-C' command line switches to execute commands but allow the continuation on to an interactive prompt if necessary. The intended use of this is to allow the starting of interactive shells with commands which occur after the user's configuration.

See individual commit messages for implementation details and decisions.

Fixes issue https://github.com/fish-shell/fish-shell/issues/4164

## TODOs:

- [x] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: src/fish.cpp [hunk 1]
```diff
@@ -214,10 +222,23 @@ static int read_init(const struct config_paths_t &paths) {
...
+        res = parser.eval(cmd_wcs, io, TOP);
+    }
+
+    return res;
+}
+
 /// Parse the argument list, return the index of the first non-flag arguments.
-static int fish_parse_opt(int argc, char **argv, std::vector<std::string> *cmds) {
-    static const char *short_opts = "+hilnvc:p:d:D:";
+static int fish_parse_opt(int argc, char **argv, command_line_switches_t *switches) {
```
Comment (krader1961) [inline src/fish.cpp]: The class and var name don't accurately convey their purpose. How about `commands_to_run_t` and `cmds`?  Alternatively, to reflect the intent to fold the other flags into this structure and be consistent with the pattern used for internal commands it could be `fish_cmd_opts_t` and `opts`.

File: src/fish.cpp [hunk 2]
```diff
@@ -57,6 +57,14 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 #define PATH_MAX 4096
 #endif
 
+// container to hold the options specified within the command line
+class command_line_switches_t {
+public:
+    std::vector<std::string> cmds;              // commands to be executed in place of interactive shell
```
Comment (krader1961) [inline src/fish.cpp]: Maybe `batch_cmds` and `postconfig_cmds`? Not essential but the use of a descriptive adverb on one member but not the other caused me to double-check my comprehension (mostly elsewhere in the code).

File: src/fish.cpp (hunk 1 above)
Comment (gerph) [inline src/fish.cpp]: I like the latter, ta.

File: src/fish.cpp (hunk 2 above)
Comment (gerph) [inline src/fish.cpp]: heh, no problem; I mostly just lifted the variable name and added the new one, so better names are appreciated :-)

**Comment by krader1961** [comment]
This looks really good. We especially appreciate the effort you made to write good unit tests. I'll test this on my macOS server later today and skim the code one more time but I'll be surprised if I find any problems based on my first read of the code.

**Comment by gerph** [comment]
Thanks - there was a failure in the macOS travis build for the -Z switch (ie a switch error), but it was hidden by the fact that `colordiff` isn't present on that system, so it wouldn't show what was wrong. I don't know what the issue is at this second, but I've changed the `invocation.sh` script to fall back to `diff` if the tool isn't present. Hopefully that'll show it up then.

I do hope the tests will be useful - it's fun finding the edge cases, although I did see some odd things when I was trying stuff out. Like I couldn't find any way to make it give me a translated version of the '-v' output, which I kinda wanted to see so that I could add language testing into the invocation test.

```
charles@work ~/f/fish-shell (initial-commands)> env LANGUAGE=de_DE:de ./fish -v
fish, version 2.6.0-156-g1ac2616-dirty
```
I'm not really sure if that's a real bug or I'm just not testing right.

:laughing: I also found that one of the tests will fail if you have the environment variable `foo` set to anything, which I don't think is particularly important, but amused me when I saw it.

**Comment by gerph** [comment]
Oooh, that's interesting. The macOS build failed with the same problems I had at home:

https://travis-ci.org/fish-shell/fish-shell/jobs/248499299#L1793-L1795

Basically, the history race condition failed on my system because I've got my home directory (which the source is under) mounted over NFS, and this seems to make those tests take a lot longer. Eventually it times out and the result is a failed test. Manually hacking things to make the test run in /tmp instead appeared to fix the problem for me, but it's interesting that it happened in the automated build too.

On my local test on my macbook, I see a failure of the 'bad switch' test. Apparently the output format is slightly different for the invalid option:

* `./fish` is reduced to `fish`
* `'Z'` as the command line option, has no quotes around it; `Z`.

```
Testing shell invocation funtionality
Testing file invocation/bad-switch.invoke ... fail
    -Z
Error output differs for file invocation/bad-switch.invoke. Diff follows:
--- invocation/bad-switch.tmp.err	2017-06-29 21:31:44.000000000 +0100
+++ invocation/bad-switch.err	2017-06-29 21:31:15.000000000 +0100
@@ -1,2 +1,2 @@
-fish: invalid option -- Z
+./fish: invalid option -- 'Z'
 RC: 1
```

Since that is the domain of the getopt library, I'm not entirely sure of the best way to handle it.
I'm going to 'fix' the problem by having a platform specific file for such a special case. I hope that's acceptable.

**Comment by krader1961** [comment]
> Since that is the domain of the getopt library, I'm not entirely sure of the best way to handle it.

The right thing to do is set `opterr` to zero to suppress the error message reported by `getopt_long()` and have the fish flag parsing loop emit an appropriate message. Our private `wgetopt_long()` does that implicitly. See any of the builtin commands such as *src/src/builtin_bind.cpp*.

**Comment by krader1961** [comment]
Another, perhaps better, solution is to run all the `argv` strings through `str2wcstring()` then use the fish private `wgetopt_long()` method since it provides consistent behavior regardless of whether we're on BSD, GNU, or something else like Haiku.

**Comment by krader1961** [comment]
> I also found that one of the tests will fail if you have the environment variable foo set to anything, which I don't think is particularly important, but amused me when I saw it.

Agreed. Don't worry about it. There are an infinite number of ways that the user environment can cause problems for unit tests. There is only so much we can do to make them hermetic when not running on a clean install like you get from Travis.

**Comment by gerph** [comment]
Do you want the getopt_long changes you suggested making here, or would that be better as a separate PR? I'm wary of changing too much beyond the main feature :-)

**Comment by krader1961** [comment]
Let's leave the getopt_long change for another PR. As you say this one is already getting somewhat unwieldy and that refinement isn't essential. I'll take another look at this and test it before going to bed. Assuming I don't find any problems I'll merge it too.

**Comment by krader1961** [comment]
Merged with some minor style cleanups (line length) and squashing two related changes. Many thanks for being so thorough and receptive to feedback, @gerph. We would love to see more changes from you given the quality of this change and your willingness to tackle modifying the core C++ code.
